### DISCUSSION TITLE
Add setting for bypassing branch policies for PRs in Azure DevOps

### DISCUSSION THREAD

**Discussion by markhakansson** [renovatebot/renovate#27755] [Suggest an Idea]
### Tell us more.

__EDIT:__ I misunderstood how this bypassing works. This is probably a bad idea to implement.

### Problem
When `automerge` is active, renovate cannot automerge PRs if branch policies (e.g., required reviewer approvals etc.) are active for the target branch. Even if the bot is allowed to bypass such restrictions.

### Suggestion
Users on Azure DevOps can be given the permission to bypass branch policies on pull requests. However, this is not enabled "automatically". It requires some additional flags to be set in the PR request when using the API.

The Azure DevOps API library that renovate uses, has support for setting this flag as an option when creating a PR. Right now, it does not.

Maybe there could be a setting e.g., `bypassBranchPolicies` or something for `renovate.json` to enable this.

### Why?
For some updates and after the build is successful for a PR, I want renovate to automerge them without any manual interventions.
However, I don't want to remove the current restrictions for branch policies for regular users. 

Which is the workaround I would have to take otherwise, and something that I've seen done in the discussions here.

### Additional information
* If build validation is enabled and branch policies are bypassable, a failing build still cannot be merged.
* Adding this would solve some discussions here related to the same topic!

**Comment by rarkins** [comment]
Related: https://github.com/renovatebot/renovate/discussions/25915

Would this be part of the "Create PR" API all, or some type of "Update PR" call after it?

**Reply by markhakansson** [reply]
For the Azure API calls, it should be possible to add this flag to either of them. It's part of the object associated with a PR that is passed to both calls.

However, in renovate the `automerge` flag (for Azure at least) is only checked when it creates a new PR. Maybe a `bypassBranchPolicies` setting should also be checked there as well?

**Reply by rarkins** [reply]
Does this concept apply to both platformAutomerge (Default behavior) as well as Renovate's own automerge?

We also need to know which specific API endpoint can be used for this, and if part of PR creation call or needs to be separate.

**Reply by markhakansson** [reply]
> Does this concept apply to both platformAutomerge (Default behavior) as well as Renovate's own automerge?

It can apply to both if we want to.

However, I think it would be safer to only allow it if `platformAutomerge: true`. I don't see the use case for allowing it without it. 

Unless one would want the bot to bypass branch policies and merge the PR even if something should go wrong, such as builds failing in the pipeline? Given that the bot (or rather the user that initiates the renovate job on Azure) has been given permission on Azure to bypass _both_ build validations and branch policies (which are separate permissions in Azure).

> We also need to know which specific API endpoint can be used for this, and if part of PR creation call or needs to be separate.

There are two main API calls used by renovate, to create and update a PR to Azure:
- [createPullRequest(...)](https://github.com/microsoft/azure-devops-node-api/blob/master/api/GitApi.ts#L5153)
- [updatePullRequest(...)](https://github.com/microsoft/azure-devops-node-api/blob/master/api/GitApi.ts#L5335)

The following object is passed to both of those calls, which contains all settings/options for the pull request:
- [GitPullRequest](https://github.com/microsoft/azure-devops-node-api/blob/a31bfaeecb2979f2c8bc123f143b964644eb02a8/api/interfaces/GitInterfaces.ts#L1683)

In `GitPullRequest.completionOptions` is where the setting to allow bypassing branch policies can be set. Thus it can be set in both calls, since the [PlatformPrOoptions](https://github.com/renovatebot/renovate/blob/049c59c09667722667df0bf1a703a971cafa997b/lib/modules/platform/types.ts#L98) are available in all instances where the API calls are made. 

Which I guess (correct me if I'm wrong!) could be useful if the configuration for renovate is updated, when the PR was previously created? Then the PR can be updated accordingly next time renovate is executing.

**Reply by rarkins** [reply]
Renovate's own automerge does not merge unless all status checks are passing. So any bypassing would not be of "tests must pass" but something else. How would it be different to when platformAutomerge is enabled and policies bypassed? e.g. does it also mean "don't worry about failing tests, merge it anyway!"?

**Reply by markhakansson** [reply]
Sorry, misread the documentation about platform automerge...

...and on a further inspection, it seems I am also wrong about how this permission works. I assumed the build validation was separate from branch policies, it is not. There is no support on Azure for bypassing _some_ branch policies. It's either all or nothing. This discussion can probably be closed then.

Thank you for your comments, really appreciate it. I hope I did not waste too much of your time on this. At least I learnt some of the inner workings of renovate, and also learnt that I need to learn how to read properly....
